sycl/include/CL/sycl/detail/spirv.hpp

```diff
@@ -41,47 +67,133 @@ template <typename Group> bool GroupAny(bool pred) {
   return __spirv_GroupAny(group_scope<Group>::value, pred);
 }
 
+// Native broadcasts map directly to a SPIR-V GroupBroadcast intrinsic
+template <typename T, typename IdT = size_t>
+using EnableIfNativeBroadcast = detail::enable_if_t<
+    detail::is_arithmetic<T>::value && std::is_integral<IdT>::value, T>;
+
+// Bitcast broadcasts can be implemented using a single SPIR-V GroupBroadcast
+// intrinsic, but require type-punning via an appropriate integer type
+template <typename T, typename IdT = size_t>
+using EnableIfBitcastBroadcast =
+    detail::enable_if_t<!detail::is_arithmetic<T>::value &&
+                            (std::is_trivially_copyable<T>::value &&
+                             (sizeof(T) == 1 || sizeof(T) == 2 ||
+                              sizeof(T) == 4 || sizeof(T) == 8)) &&
+                            std::is_integral<IdT>::value,
+                        T>;
+
+template <typename T>
+using ConvertToNativeBroadcastType_t = select_cl_scalar_integral_unsigned_t<T>;
+
+// Generic broadcasts may require multiple calls to SPIR-V GroupBroadcast
+// intrinsics, and should use the fewest broadcasts possible
+// - Loop over 64-bit chunks until remaining bytes < 64-bit
+// - At most one 32-bit, 16-bit and 8-bit chunk left over
+template <typename T, typename IdT = size_t>
+using EnableIfGenericBroadcast =
```

Comment:
I do not see difference between this and EnableIfBitcastBroadcast. is it expected?

Reply:
There's an extra `!` in front of the second condition, which I agree is hard to see.  I'll have a think and see if there's a simpler way to express these conditions.

Reply:
I've tried to tidy things up a little bit in 0bdb196.  Is it clearer?

Reply:
yes. thank you